filebeat/input/filestream/internal/task/group.go

```diff
@@ -0,0 +1,129 @@
+package task
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type Group struct {
+	limit uint64
+
+	sem *semaphore.Weighted
+	wg  *sync.WaitGroup
+
+	errs     []error
+	errsSize uint64
+	errsNext uint64
+	errsMu   *sync.Mutex
```

Comment:
Looking at some more of the code that uses `errs` and it's related fields, I'm wondering if it might be a bit cleaner to create a more general-purpose ring buffer type and use it for `errs` here.  That type could provide an `Add(T)` method that could be called from `addErr` and an `IsEmpty() bool` method that could be called from `Stop()`.

Reply:
Based on our conversation off-PR, you told me you're getting rid of the idea of the ring buffer for errors altogether.  So this thread is now irrelevant.